src/ghci/stdout.rs

```diff
@@ -36,12 +39,56 @@ pub struct GhciStdout {
     pub stdin_sender: mpsc::Sender<StdinEvent>,
     pub stderr_sender: mpsc::Sender<StderrEvent>,
     pub receiver: mpsc::Receiver<StdoutEvent>,
+    pub prompt_patterns: AhoCorasick,
     pub buffer: Vec<u8>,
 }
 
 impl GhciStdout {
     #[instrument(skip_all, name = "stdout", level = "debug")]
-    pub async fn run(mut self, when_ready: oneshot::Sender<()>) -> miette::Result<()> {
+    pub async fn run(mut self) -> miette::Result<()> {
+        let mut backoff = ExponentialBackoff::default();
+        while let Some(duration) = backoff.next_backoff() {
+            match self.run_inner().await {
+                Ok(()) => {
+                    // MPSC channel closed, probably a graceful shutdown?
+                    tracing::debug!("Channel closed");
+                    break;
+                }
+                Err(err) => {
+                    tracing::error!("{err:?}");
+                }
+            }
+
+            tracing::debug!("Waiting {duration:?} before retrying");
+            tokio::time::sleep(duration).await;
+        }
+
+        Ok(())
+    }
+
+    async fn run_inner(&mut self) -> miette::Result<()> {
```

Comment:
This isn't new, I just moved a bunch of the startup logic into the `initialize` method and it confused the diff.